### PR TITLE
Integrates GitLab Runner into DTaaS

### PR DESCRIPTION
**Pull request by aryanpingle**
# Integrates GitLab Runner into DTaaS

Addresses #892 by integrating a GitLab Runner setup into the DTaaS software. 

## Type of Change

✔ New feature
✔ Documentation update

## Description

Adds a `compose.runner.yml` file that spins up a Docker container running the `gitlab/gitlab-runner` image, and automatically starts a gitlab runner based on the settings in `config.toml`. 

## Testing

Tested using a localhost and a server installation, both using a server-hosted gitlab instance.

## Impact

The process of registering a GitLab Runner will be greatly simplified. Developers will be able to use the digital twins preview page with a self-hosted runner.

## Additional Information

- [ ] On localhost, the hello-world digital twin fails during execution as the runner executor cannot access the host machine's localhost.
- [ ] Currently, `network_mode` is set to `host` for the gitlab-runner container. This should instead use or create a docker network.
- [X] Added a guide in the `services/gitlab` directory.
- [ ] Need to update/create the admin guide for using the Digital Twins Preview Page.

## Checklist

- [X] My code adheres to the coding and style guidelines of the project.
- [X] I have added tests for all the new code and any changes made to
      existing code.
- [X] I have made corresponding changes to the documentation.

### CODE REVIEW

**Comment by aryanpingle** [comment]
The `gitlab-runner` container is built on the `gitlab/gitlab-runner` image using alpine and contains the `gitlab-runner` executable. Starting the container executes `gitlab-runner register` and `gitlab-runner start` automatically, so we only need to deal with the `runner-config.toml` file to set things up.

The `gitlab/gitlab-runner` image uses a custom image of its own, from the "registry.gitlab.com/gitlab-org/gitlab-runner/gitlab-runner-helper" repository. This uses the docker engine to spawn a **sibling** container to `gitlab-runner`, which actually executes the CI/CD job.

Now, if you try using `docker ps` you won't see the sibling executor container/s, for a reason I don't fully understand. Here's a link to the SO answer: [https://stackoverflow.com/a/41994792/8089674](https://stackoverflow.com/a/41994792/8089674).

**Comment by prasadtalasila** [comment]
@aryanpingle A related correction in the docs

line-69 of **deploy/services/gitlab/INTEGRATION.md** file should be:

```txt
If the DTaaS application is hosted at https://foo.com/, then configure the following files:
```

File: deploy/services/runner/GITLAB-RUNNER.md
Comment (prasadtalasila) [inline deploy/services/runner/GITLAB-RUNNER.md:1]: @aryanpingle please keep documentation of both instance and project runners

File: deploy/services/runner/GITLAB-RUNNER.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+and executes jobs for a single project, or any project within a GitLab group.
+
+This document outlines the steps needed to create a Docker container named
+`gitlab-runner` which will contain a single runner that will be responsible for
+the execution of Digital Twins. There are two installation scenarios:
+
+1. Localhost Installation - You are using the integrated runner locally with
+   a GitLab instance hosted at `https://localhost/gitlab`.
+2. Server Installation - You are using the integrated runner with a GitLab
+   instance hosted on a production server. This server may be a remote server
```
Comment (prasadtalasila) [inline deploy/services/runner/GITLAB-RUNNER.md:10]: (`https://localhost/gitlab`)

File: deploy/services/gitlab/GITLAB-RUNNER.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+In this directory, the file `gitlab-runner/config.toml` should have the
+following contents:
+
+```toml
+[[runners]]
+  name = "dtaas-runner-0"
+  url = "https://foo.com/gitlab/"
+  token = "xxx"
+  executor = "docker"
+  [runners.docker]
```
[the remark this answers is not in the document]
Comment (prasadtalasila) [reply deploy/services/gitlab/GITLAB-RUNNER.md]: yes

**Comment by prasadtalasila** [comment]
> The `gitlab-runner` container is built on the `gitlab/gitlab-runner` image using alpine and contains the `gitlab-runner` executable. Starting the container executes `gitlab-runner register` and `gitlab-runner start` automatically, so we only need to deal with the `runner-config.toml` file to set things up.
> 
> The `gitlab/gitlab-runner` image uses a custom image of its own, from the "registry.gitlab.com/gitlab-org/gitlab-runner/gitlab-runner-helper" repository. This uses the docker engine to spawn a **sibling** container to `gitlab-runner`, which actually executes the CI/CD job.
> 
> Now, if you try using `docker ps` you won't see the sibling executor container/s, for a reason I don't fully understand. Here's a link to the SO answer: https://stackoverflow.com/a/41994792/8089674.

please add this information in a separate section at the end of GITLAB-RUNNER.md

File: docs/developer/devops/api.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# API Calls
+
+A pipeline can be triggered via an API call using a
```
Comment (prasadtalasila) [inline docs/developer/devops/api.md]: Gitlab DevOps pipeline....

File: docs/developer/devops/api.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+[pipeline trigger token](https://docs.gitlab.com/ee/ci/triggers/)
+which is created on the GitLab instance, with the following values:
+
+1. `<access_token>`: The user GitLab trigger token.
```
Comment (prasadtalasila) [inline docs/developer/devops/api.md]: How about just `<trigger_token>`?

**Comment by aryanpingle** [comment]
The latest commits add content from Chapter 3, Section 3 and Chapter 4 from "Design and implementation of an integrated DevOps framework for Digital Twins as a Service software platform" (Scherma V.) as a new section "DevOps Framework" in the developer documentation. I've reordered a few sections and edited some parts to make it more developer friendly.

File: docs/developer/devops/gitlab-ci-cd.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
+# GitLab CI/CD Infrastructure
+
+Given that files in the Library are stored in a Git repository, the approach
```
Comment (prasadtalasila) [inline docs/developer/devops/gitlab-ci-cd.md:3]: Gitlab parent-child pipeline (with gitlab.com....ref) are used to trigger execution of digital twins stored in one gitlab repository.

File: docs/developer/devops/implementation.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
+# Implemented Classes
+
+In order to facilitate the management of the lifecycle of DTs via the web
+application interfaces, it was necessary to develop specific code within the
+project client. The code was designed to facilitate efficient API calls through
+the use of Gitbeaker as a wrapper, as this approach simplifies interactions
+with GitLab’s REST API and reduces the complexity of the project code.
+
+After testing the correctness of the API by implementing a `GitlabDriver`
```
Comment (prasadtalasila) [inline docs/developer/devops/implementation.md]: This file has been removed. Please rephrase these sentences.

File: docs/developer/devops/implementation.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+latter is monitored so that, once the entire pipeline is complete, the results
+can be displayed in detail within the user interface. In this way, all statuses
+of each operation are logged for better debugging and performance analysis,
+including possible errors. Having trace logs exposed to the user means
+troubleshooting will be more effective and insight into execution and
+management of DTs will be gained, improving system reliability and user
+confidence.
+
+## DigitalTwin
+
```
Comment (prasadtalasila) [inline docs/developer/devops/implementation.md:109]: Please put a typescript code block containing class name and its methods. You can copy the class name and method signatures from codebase. The explanation makes sense w.r.t this diagram.

Pæease do this for all the classes explained in this PR.

**Review by prasadtalasila — COMMENTED** [review]
@aryanpingle thanks for the updates. please see the comments.

**Comment by prasadtalasila** [comment]
@aryanpingle please add the following citation at the end of pages taken from Vanessa's thesis:

```txt
Ref: Vanessa Scherma, Design and implementation of an integrated DevOps framework for Digital Twins as a Service software platform, Master’s Degree Thesis, Politecnico Di Torino, 2024.
```

**Comment by prasadtalasila** [comment]
Thanks @aryanpingle. The changes have been merged via PR #1115.

**Pull request closed without merging**